### PR TITLE
[VidLii] Add HD support

### PR DESCRIPTION
**Pull request by mrpapersonic**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

VidLii has a [partnership program](https://www.vidlii.com/partners) that allows users to have higher quality videos. This pull request checks for the higher quality source and adds it to the formats list.

**Example videos**
- [FULPTUBE SUCKS.](https://www.vidlii.com/watch?v=zTAtaAgOLKt)

### CODE REVIEW

**Review by brighthuo — APPROVED** [review]
Ok

File: youtube_dl/extractor/vidlii.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 from __future__ import unicode_literals
 
 import re
+import requests
```
Comment (dstftw) [inline youtube_dl/extractor/vidlii.py:5]: Do not use `requests`.

File: youtube_dl/extractor/vidlii.py [hunk 2]
```diff
@@ -17,23 +18,23 @@
...
+            'thumbnail': 'https://www.vidlii.com/usfi/thmp/zTAtaAgOLKt.jpg',
+            'uploader': 'Homicide',
+            'uploader_url': 'https://www.vidlii.com/user/Homicide',
+            'upload_date': '20210612',
+            'duration': 89,
             'view_count': int,
             'comment_count': int,
             'average_rating': float,
             'categories': ['News & Politics'],
-            'tags': ['Vidlii', 'Jan', 'Videogames'],
```
Comment (dstftw) [inline youtube_dl/extractor/vidlii.py:36]: Do not remove existing tests.

File: youtube_dl/extractor/vidlii.py [hunk 3]
```diff
@@ -42,17 +43,34 @@ class VidLiiIE(InfoExtractor):
...
+            })
+
+        hdsrc = self._search_regex(
+            r'hdsrc\s*:\s*(["\'])(?P<url>(?:https?://)?(?:(?!\1).)+)\1',
+            webpage, 'video url', group='url')
+        if not requests.head(hdsrc).status_code == 404:
+            add_format(hdsrc)
+        add_format(self._search_regex(
+            r'src\s*:\s*(["\'])(?P<url>(?:https?://)?(?:(?!\1).)+)\1',
+            webpage, 'video url', group='url'))
```
Comment (dstftw) [inline youtube_dl/extractor/vidlii.py:68]: DRY: 61-63, 66-68.

File: youtube_dl/extractor/vidlii.py [hunk 4]
```diff
@@ -42,17 +43,34 @@ class VidLiiIE(InfoExtractor):
...
+        self._sort_formats(formats)
 
         title = self._search_regex(
-            (r'<h1>([^<]+)</h1>', r'<title>([^<]+) - VidLii<'), webpage,
-            'title')
+            (r'<h1>([^<]+)</h1>', r'<title>([^<]+) - VidLii<'),
+            webpage, 'title')
```
Comment (dstftw) [inline youtube_dl/extractor/vidlii.py:73]: Avoid unrelated changes.

File: youtube_dl/extractor/vidlii.py [hunk 5]
```diff
@@ -42,17 +43,34 @@ class VidLiiIE(InfoExtractor):
...
-        video_url = self._search_regex(
-            r'src\s*:\s*(["\'])(?P<url>(?:https?://)?(?:(?!\1).)+)\1', webpage,
-            'video url', group='url')
+        formats = []
+
+        def add_format(format_url, height=None):
+            if format_url[-7:][:-4] == "720":
+                height = 720
+            else:
+                height = 360
```
Comment (dstftw) [inline youtube_dl/extractor/vidlii.py:54]: Parse height with regex instead.

**Review by dstftw — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/vidlii.py (hunk 1 above)
Comment (mrpapersonic) [reply youtube_dl/extractor/vidlii.py]: Now uses the included urllib instead of requests.

File: youtube_dl/extractor/vidlii.py (hunk 2 above)
Comment (mrpapersonic) [reply youtube_dl/extractor/vidlii.py]: Readded.

File: youtube_dl/extractor/vidlii.py (hunk 5 above)
Comment (mrpapersonic) [reply youtube_dl/extractor/vidlii.py]: Now uses regex instead of the hack.

**Comment by dstftw** [comment]
Read tutorial and coding conventions.

**Comment by mrpapersonic** [comment]
@dirkf Can this be re-reviewed?

File: youtube_dl/extractor/vidlii.py [hunk 6]
```diff
@@ -36,19 +39,60 @@ class VidLiiIE(InfoExtractor):
...
+            })
+
+        hdsrc = self._search_regex(
+            r'hdsrc\s*:\s*(["\'])(?P<url>(?:https?://)?(?:(?!\1).)+)\1',
+            webpage, 'video url', group='url')
+        try:
+            compat_urllib_request.urlopen(hdsrc)
+            add_format(hdsrc)
+        except Exception:
+            pass
```
Comment (pukkandan) [inline youtube_dl/extractor/vidlii.py:91]: You should use `self._request_webpage`, preferably with a HEAD request

File: youtube_dl/extractor/vidlii.py [hunk 7]
```diff
@@ -36,19 +39,60 @@ class VidLiiIE(InfoExtractor):
...
+        add_format(self._search_regex(
+            r'src\s*:\s*(["\'])(?P<url>(?:https?://)?(?:(?!\1).)+)\1',
+            webpage, 'video url', group='url'))
```
Comment (pukkandan) [inline youtube_dl/extractor/vidlii.py:94]: this is basically the same code repeated twice. It can be generalized

File: youtube_dl/extractor/vidlii.py [hunk 8]
```diff
@@ -36,19 +39,60 @@ class VidLiiIE(InfoExtractor):
...
-            'video url', group='url')
+        formats = []
+
+        def add_format(format_url, height=None):
+            try:
+                height = int(self._search_regex(r"(?P<h>\d\d\d).mp4",
+                                                format_url, "height",
+                                                group="h"))
+            except Exception:
+                height = 360
```
Comment (pukkandan) [inline youtube_dl/extractor/vidlii.py:77]:
```suggestion
            height = int(self._search_regex(r'(\d+).mp4', format_url, 'height', default=360))
```

**Comment by dirkf** [comment]
Superseded by #30924.
